File: 백준/Gold/1806. 부분합/부분합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int N, S, result = Integer.MAX_VALUE;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            S = Integer.parseInt(st.nextToken());
            arr = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
            
            int lt = 0;
            int rt = 0;
            int sum = 0;
            int count = 0;
            while (rt < N) {
                sum += arr[rt];
                rt++;
                count++;
                if (sum >= S) {
                    result = Math.min(result, count);
                    while (sum >= S) {
                        sum -= arr[lt];
                        count--;
                        lt++;
                        if (sum >= S) {
                            result = Math.min(result, count);
                        }
                    }
                }
            }

            result = result == Integer.MAX_VALUE ? 0 : result;
            System.out.println(result);
        }
    }
}

File: 백준/Gold/1806. 부분합/부분합.kt
import java.io.BufferedReader
import java.util.StringTokenizer
import kotlin.math.min

fun main() {
    BufferedReader(System.`in`.reader()).use { br ->
        val st = StringTokenizer(br.readLine())
        val N = st.nextToken().toInt()
        val S = st.nextToken().toInt()
        val arr = br.readLine().split(" ")
            .map { it.toInt() }
            .toTypedArray();

        val answer = partialSum(N, S, arr)
        println(answer)
    }
}

private fun partialSum(N: Int, S: Int, arr: Array<Int>): Int {
    var minLength = Int.MAX_VALUE
    var length = 1
    var sum = 0
    var lt = 0
    var rt = 0
    while (rt < N) {
        sum += arr[rt]

        while (sum >= S) {
            minLength = min(minLength, length)
            sum -= arr[lt]
            lt++
            length--
        }

        rt++
        length++
    }

    return if (minLength == Int.MAX_VALUE) 0 else minLength
}